python/cudf/cudf/core/reshape.py

```diff
@@ -275,44 +277,50 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
         df = cudf.DataFrame()
         _normalize_series_and_dataframe(objs, axis=axis)
 
-        old_objs = objs
+        # Inner joins involving empty data frames always return empty dfs, but
+        # We must delay returning until we have set the column names.
+        empty_inner = any(obj.empty for obj in objs) and join == "inner"
+
         objs = [obj for obj in objs if obj.shape != (0, 0)]
+
         if len(objs) == 0:
             return df
-        empty_inner = False
-        if join == "inner":
-            # don't filter out empty df's
-            if any(obj.empty for obj in old_objs):
-                empty_inner = True
 
         objs, match_index = _align_objs(objs, how=join)
 
-        for idx, o in enumerate(objs):
-            if idx == 0:
-                df.index = o.index
-            for col in o._data.names:
-                if col in df._data:
+        df.index = objs[0].index
+        for o in objs:
+            for name, col in o._data.items():
+                if name in df._data:
                     raise NotImplementedError(
-                        f"A Column with duplicate name found: {col}, cuDF "
+                        f"A Column with duplicate name found: {name}, cuDF "
                         f"doesn't support having multiple columns with "
                         f"same names yet."
                     )
-                df[col] = o._data[col]
+                df[name] = col
+
+        result_columns = objs[0].columns.append(
+            [obj.columns for obj in objs[1:]]
+        )
 
-        result_columns = objs[0].columns
-        for o in objs[1:]:
-            result_columns = result_columns.append(o.columns)
         if ignore_index:
             # with ignore_index the column names change to numbers
             df.columns = pd.RangeIndex(len(result_columns.unique()))
         else:
             df.columns = result_columns.unique()
+
         if empty_inner:
             # if join is inner and it contains an empty df
             # we return an empty df
             return df.head(0)
+
+        # This check uses `sort is not False` rather than just `sort=True`
+        # to differentiate between a user-provided `False` value and the
+        # default `None`. This is necessary for pandas compatibility, even
+        # though `True` and `False` are the only valid options from the user.
         if not match_index and sort is not False:
```

Comment:
I think `sort` is by default `False` now? https://pandas.pydata.org/docs/reference/api/pandas.concat.html . Note if we are changing ours, it should be a different PR.

Reply:
Was the default for pandas concat `sort=True` at one point?

Reply:
Prior to v1.0 it was defaulted at `None`, but the meaning of `sort=None` is actually `sort=True`. They changed that after v1.0: https://github.com/pandas-dev/pandas/blob/13653812b43aab0d4be29177f9c849c07414a7bc/pandas/core/reshape/concat.py#L86-L89

Reply:
Sounds like we are missing a test for `None` parameter for `sort`?

Reply:
Got it. So the current behavior of `cudf` is to have `sort=None`, which means `sort=False`, so that's consistent with `pandas`. I assumed (but I don't remember whether I tested) that the line of code we're currently discussing (the `sort is not False`) is specifically to handle an edge case where `pandas` sorted the data even when `sort=False`, but I'm not 100% sure if that's the case. I'll double-check that before finalizing this PR, maybe we can simply change the behavior in this line. I know that it caused some tests to fail, but maybe those tests should be modified.

Reply:
Yes, this is necessary. I think it's specifically an issue for non-integral indexes where `cudf` won't generate a sorted index without explicitly calling `sort_index` here but pandas will sort on the index even when the user doesn't request a sort. See `test_concat.py::test_pandas_concat_compatibility_axis1`.